Add initial_color property to CalendarEntity

Misiu opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a color property to CalendarEntity.

Architecture discussion got approved: https://github.com/home-assistant/architecture/discussions/883 🎉🎉🎉

Frontend: https://github.com/home-assistant/frontend/pull/28882

Docs: https://github.com/home-assistant/developers.home-assistant/pull/2919

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 


I need to know how to change the config flow to display the actual color picker. I saw there is `ColorRGBSelector` (and color_rgb_to_hex), but I'm not sure if this is the right way to do it.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Misiu committed:
Add color attribute support to calendar entities and update config flow

Misiu committed:
Add color attribute support to calendar entities and update config flow

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -523,21 +523,29 @@ def event(self) -> CalendarEvent | None:
             "location": event.location if event.location else "",
             "description": event.description if event.description else "",
         }
+        if (color := self.color) is not None:
```
This is the discussion I was thinking of: https://github.com/home-assistant/architecture/discussions/680

Misiu replied:
Done

allenporter replied:
I have re-read the original architecture issue and maybe I want to change my stance here. Thanks for understanding the churn, but maybe we shouldn't be updating all the unrelated tests.

I think what we can say is: If the device has a color then it can always set the color attribute. If it doesn't then it doesn't, like you had it originally.

I think here we can say "supported by the entity" to mean "has a color assigned". Then we'll make sure its always set, even if an event is not present etc.

Misiu replied:
No worries :)
So, if I understand correctly, I should remove the feature and change the logic for state_attributes. I'll always return the color if it is not None (even when there are no events in the calendar).

allenporter commented:
Are you planning on adding frontend support for the colors?

I've exposed  Google Calendar color from the API in #145642 by the way.

Misiu commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -523,21 +523,29 @@ def event(self) -> CalendarEvent | None:
 
+    @property
+    def color(self) -> str | None:
+        """Return the color of the calendar entity."""
```
Thank you for the initial review. I was thinking about using hex-formatted strings, no alpha, for example `#019CDC`. I think it will be easier on the frontend.

allenporter replied:
Sure, what I mean to say is this should say what the format is, or define a type, so its not just arbitrary strings here.

I see there is a type `RGBColor` which could be used to enforce that this meets the format (e.g. integrations provide `RGBColor` and the entity can convert to hex string for frontend.

Misiu replied:
I was thinking about using ColorRGBSelector for the config/option flow, so using RGBColor for the property will make things easier. Will try that.

Misiu replied:
I've changed the color property to be RGBColor

Misiu committed:
Add RGB color attribute support to calendar entities and update tests

Misiu committed:
Add RGB color attribute support to calendar entities and update tests

Misiu commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -523,21 +523,29 @@ def event(self) -> CalendarEvent | None:
+    @property
+    def color(self) -> str | None:
+        """Return the color of the calendar entity."""
+        return None
```
Done

Misiu committed:
use ColorRGBSelector in config flow

Misiu committed:
use ColorRGBSelector in config flow

Misiu commented:
> Are you planning on adding frontend support for the colors?

I plan to make all the backend changes, and when they are approved, I'll try to make the frontend changes.

For now, I removed the tests for the color property. I'll try to add option flow first, and then, when you are ok with the changes, I'll add tests.

Misiu committed:
fix remote_calendar and holidays tests

Misiu committed:
fix remote_calendar and holidays tests

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -529,14 +537,16 @@ def state_attributes(self) -> dict[str, Any] | None:
-            "description": event.description if event.description else "",
+            "location": event.location if event.location is not None else "",
+            "description": event.description if event.description is not None else "",
+            "color": f"#{color_rgb_to_hex(*self._attr_color)}"
+            if self._attr_color is not None
+            else None,
```
I'm not sure this meets the style guide so i think it may need to be done as a variable

Misiu replied:
I used a variable, the same way as in home_connect

allenporter commented on homeassistant/components/local_calendar/calendar.py:
```diff
@@ -84,12 +89,22 @@ def __init__(
         self._event: CalendarEvent | None = None
         self._attr_name = name
         self._attr_unique_id = unique_id
+        self._attr_color: RGBColor | None = color
```
I think we can just write as:
```suggestion
        self._attr_color = color
```

allenporter commented on homeassistant/components/local_calendar/calendar.py:
```diff
@@ -84,12 +89,22 @@ def __init__(
         """Return the next upcoming event."""
         return self._event
 
+    @property
+    def color(self) -> RGBColor | None:
+        """Return the color of the calendar entity as RGBColor."""
+        return self._attr_color
+
+    @color.setter
+    def color(self, value: RGBColor | None) -> None:
+        self._attr_color = value
+
```
I don't think we need this
```suggestion
```

allenporter reviewed (changes requested):
Seems like this is going in a good direction. I'd still like to see this not as part of the local calendar config flow, but as part of the options flow.

Remove the setters, i don't think we want those.

Misiu committed:
Update homeassistant/components/local_calendar/calendar.py

Co-authored-by: Allen Porter <allen.porter@gmail.com>

Misiu committed:
Update homeassistant/components/local_calendar/calendar.py

Co-authored-by: Allen Porter <allen.porter@gmail.com>

Misiu committed:
Update homeassistant/components/local_calendar/calendar.py

Co-authored-by: Allen Porter <allen.porter@gmail.com>

Misiu committed:
Update homeassistant/components/local_calendar/calendar.py

Co-authored-by: Allen Porter <allen.porter@gmail.com>

Misiu committed:
use variable for hex color

Misiu committed:
use variable for hex color

Misiu committed:
options flow

Misiu committed:
options flow

Misiu committed:
Refactor options flow handler and add tests for Local Calendar options flow

Misiu committed:
Refactor options flow handler and add tests for Local Calendar options flow

Misiu commented:
New config flow for Local Calendar:
![image](https://github.com/user-attachments/assets/3ca61d9a-8ae3-4366-937a-bb263f0f4441)

Options flow:
![image](https://github.com/user-attachments/assets/1dd56984-a663-4c40-9363-5a7deedfe36b)

States:
![image](https://github.com/user-attachments/assets/f9099763-7952-469b-9ff0-3402f2c01c1c)
The second calendar doesn't have any events, so its color isn't available in attributes.

Misiu committed:
Add test for calendar color attribute exposure in state attributes

Misiu committed:
Add test for calendar color attribute exposure in state attributes

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -530,13 +538,17 @@ def state_attributes(self) -> dict[str, Any] | None:
         if (event := self.event) is None:
```
I realize from yoru comment/screenshot: We shouldn't stop exporting the color because there are no upcoming events. It needs to be always present.

Misiu replied:
done

Misiu committed:
Add color support to calendar entities and update tests accordingly

Misiu committed:
Add color support to calendar entities and update tests accordingly

Misiu commented:
I've made the requested changes, added a new feature, and added it to the local calendar. I've also changed how state_attributes are returned. Now, color is returned if SUPPORTS_COLOR is added to integration, and the color is returned even if there are no events.

![image](https://github.com/user-attachments/assets/3f0f7a7c-4688-4c31-a9fe-811fc2efe23a)

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -518,6 +519,13 @@ class CalendarEntity(Entity):
+    _attr_color: RGBColor | None = None
+
+    @property
+    def color(self) -> RGBColor | None:
```
This is currently untested. I think the `_attr_color` function needs to call `self.color` and not get the attribute directly in case the integration overrides the color property and doesn't use the attribute.

allenporter replied:
In `state_attributes` call `self.color` instead of `self._attr_color`

Misiu replied:
done

NoRi2909 commented on homeassistant/components/local_calendar/strings.json:
```diff
@@ -6,6 +6,7 @@
         "description": "Please choose a name for your new calendar",
         "data": {
           "calendar_name": "Calendar Name",
+          "calendar_color": "Calendar Color",
           "import": "Starting Data"
```
Those three all need to be changed to sentence-casing:

          "calendar_name": "Calendar name",
          "calendar_color": "Calendar color",
          "import": "Starting data"

We can do the two existing ones in a separate PR or just include the fix here.

Misiu replied:
done

NoRi2909 commented on homeassistant/components/local_calendar/strings.json:
```diff
@@ -27,5 +28,14 @@
+    "step": {
+      "init": {
+        "data": {
+          "calendar_color": "Calendar Color"
```
This can be replaced with a reference to the one in line 9 which will also fix the sentence-casing.

Misiu replied:
How should the key look?
```[%key:component::local_calendar::config::step::user::data::calendar_color%]```

NoRi2909 replied:
This looks OK to me.

Misiu replied:
fixed

Misiu committed:
Update strings.json

Misiu committed:
Update strings.json

Misiu committed:
Update strings.json for improved calendar configuration

Misiu committed:
Update strings.json for improved calendar configuration

Misiu committed:
Remove SUPPORTS_COLOR and fix tests

Misiu committed:
Remove SUPPORTS_COLOR and fix tests

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -527,18 +535,28 @@ def event(self) -> CalendarEvent | None:
-            return None
+            if color_hex_val is None:
+                return None
+            return {"color": color_hex_val}
```
Please add a test for a case where there is no upcoming event but the calendar has a color.

You may just want to add another MockCalendarEntity to the list of entities to exercise this case. You can also pass a color in the constructor instead of changing the attributes in the middle of the test.

Misiu committed:
use self.color instead of self._attr_color

Misiu committed:
use self.color instead of self._attr_color

Misiu committed:
test calendar with color and no events

Misiu committed:
test calendar with color and no events

Misiu committed:
all green

Misiu committed:
all green

allenporter reviewed (changes requested):
Tests look great, thank you. Some comments below.

allenporter commented:
I just realized the RFC has support for a `COLOR` attribute at the calendar level:

https://datatracker.ietf.org/doc/html/rfc7986#section-4

so we may want to consider in the case of local calendar storing the color in the calendar so that it can be exported as well.

allenporter commented:
> I just realized the RFC has support for a `COLOR` attribute at the calendar level:
> 
> https://datatracker.ietf.org/doc/html/rfc7986#section-4
> 
> so we may want to consider in the case of local calendar storing the color in the calendar so that it can be exported as well. But this also means we can read the color from uploaded calendars or read it from remote calendar as well.

I would like to go back to the arch discussion and clarify a few things before we proceed. Sorry for the hold up here, i realize some of the decision before is ambiguous to me.

Misiu commented:
Still actual

Misiu committed:
Use entity registry options for calendar color

Store calendar color in entity registry options instead of state
attributes. This follows the pattern established by the lock entity
with default_code (PR #88139).

Changes:
- Add CONF_COLOR constant to calendar domain
- Add entity option reading to CalendarEntity via
  async_internal_added_to_hass, async_registry_entry_updated, and
  _async_read_entity_options
- Add get_initial_entity_options to LocalCalendarEntity for migration
  from config entry to entity registry options
- Remove color from state_attributes in CalendarEntity
- Remove options flow from local_calendar (color managed via entity
  registry instead)
- Update tests to verify new behavior

Misiu committed:
Use entity registry options for calendar color

Store calendar color in entity registry options instead of state
attributes. This follows the pattern established by the lock entity
with default_code (PR #88139).

Changes:
- Add CONF_COLOR constant to calendar domain
- Add entity option reading to CalendarEntity via
  async_internal_added_to_hass, async_registry_entry_updated, and
  _async_read_entity_options
- Add get_initial_entity_options to LocalCalendarEntity for migration
  from config entry to entity registry options
- Remove color from state_attributes in CalendarEntity
- Remove options flow from local_calendar (color managed via entity
  registry instead)
- Update tests to verify new behavior

Misiu committed:
Merge branch 'dev' into color

Misiu committed:
Improve calendar color entity registry tests

- Add unique_id support to MockCalendarEntity for entity registry
- Add test_calendar_color_from_entity_options to test entity registry
  options functionality (setting/clearing color via registry)
- Add test_calendar_color_fallback_to_attr to test precedence:
  registry options take precedence over _attr_color
- Update create_test_entities to assign unique IDs to test entities

Misiu committed:
Improve calendar color entity registry tests

- Add unique_id support to MockCalendarEntity for entity registry
- Add test_calendar_color_from_entity_options to test entity registry
  options functionality (setting/clearing color via registry)
- Add test_calendar_color_fallback_to_attr to test precedence:
  registry options take precedence over _attr_color
- Update create_test_entities to assign unique IDs to test entities

Misiu committed:
Merge branch 'dev' into color

Misiu commented:
@allenporter I've done requested changes and also created frontend PR: https://github.com/home-assistant/frontend/pull/28882
I'm sure it's not the best, but it works fine. I've tested it with calendar panel and card.

Looking forward to a review.

Misiu committed:
Merge branch 'dev' into color

Misiu committed:
Merge branch 'dev' into color

Misiu committed:
remove config flow, we store color in entity options

Misiu committed:
remove config flow, we store color in entity options

Misiu committed:
revert formatting to minimize the changes

Misiu committed:
revert formatting to minimize the changes

Misiu committed:
remove unused update listener from Local Calendar setup

Misiu committed:
remove unused update listener from Local Calendar setup

Misiu committed:
Merge branch 'dev' into color

allenporter reviewed (changes requested):
Great job moving this forward.

Misiu committed:
Merge branch 'dev' into color

Misiu committed:
Refactor test to use public API instead of internal attributes

Misiu committed:
Refactor test to use public API instead of internal attributes

Misiu committed:
don't use assert

Misiu committed:
don't use assert

Misiu commented on homeassistant/components/local_calendar/calendar.py:
```diff
@@ -56,7 +57,12 @@ async def async_setup_entry(
     name = config_entry.data[CONF_CALENDAR_NAME]
-    entity = LocalCalendarEntity(store, calendar, name, unique_id=config_entry.entry_id)
+    color = config_entry.options.get(
+        CONF_CALENDAR_COLOR, config_entry.data.get(CONF_CALENDAR_COLOR)
```
This should be fixed

Misiu committed:
Refactor calendar entity color handling and update tests for initial color validation

Misiu committed:
Refactor calendar entity color handling and update tests for initial color validation

Misiu committed:
Merge branch 'color' of https://github.com/Misiu/home-assistant into color

Misiu committed:
Merge remote-tracking branch 'upstream/dev' into color

Misiu committed:
Merge branch 'dev' into color

Misiu committed:
prek

Misiu committed:
prek

Misiu committed:
Merge branch 'dev' into color

Misiu committed:
Rename color attribute to initial_color in MockCalendarEntity and update related tests

allenporter reviewed (commented):
Overall i think this looks great and should be ready to go except a question and a few small comments on tests

allenporter reviewed (changes requested):
(requested changes in other comment)

Misiu committed:
Add initial_color property

Misiu committed:
review changes

Misiu committed:
Merge branch 'dev' into color

Misiu committed:
fix prek

allenporter reviewed (approved):
Looks great. We should now focus on approving the frontend PR.

Misiu committed:
Merge branch 'dev' into color
